File: intrinsic/assets/dependencies/utils_test.go
// Copyright 2023 Intrinsic Innovation LLC

package utils

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"testing"

	"intrinsic/assets/data/fakedataassets"
	"intrinsic/testing/grpctest"
	"intrinsic/util/proto/descriptor"
	"intrinsic/util/proto/testing/prototestutil"

	"github.com/google/go-cmp/cmp"
	"google.golang.org/grpc"
	"google.golang.org/grpc/metadata"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/testing/protocmp"

	dapb "intrinsic/assets/data/proto/v1/data_asset_go_proto"
	tcpb "intrinsic/assets/dependencies/testing/test_configs_go_proto"
	tsgrpcpb "intrinsic/assets/dependencies/testing/test_service_go_grpc_proto"
	tspb "intrinsic/assets/dependencies/testing/test_service_go_grpc_proto"
	atpb "intrinsic/assets/proto/asset_type_go_proto"
	idpb "intrinsic/assets/proto/id_go_proto"
	mpb "intrinsic/assets/proto/metadata_go_proto"
	gcpb "intrinsic/assets/proto/v1/grpc_connection_go_proto"
	rdpb "intrinsic/assets/proto/v1/resolved_dependency_go_proto"

	anypb "google.golang.org/protobuf/types/known/anypb"
	emptypb "google.golang.org/protobuf/types/known/emptypb"
)

type testService struct{}

func (s *testService) Test(ctx context.Context, req *tspb.TestRequest) (*tspb.TestResponse, error) {
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		return nil, fmt.Errorf("no metadata found in context")
	}

	response := &tspb.TestResponse{
		ContextMetadata: make(map[string]*tspb.TestResponse_MetadataValues),
	}
	for k, vs := range md {
		response.GetContextMetadata()[k] = &tspb.TestResponse_MetadataValues{
			Values: vs,
		}
	}

	return response, nil
}

func newTestServer(t *testing.T) string {
	t.Helper()

	var s *testService
	grpcServer := grpc.NewServer()
	tsgrpcpb.RegisterTestServiceServer(grpcServer, s)

	return grpctest.StartServerT(t, grpcServer)
}

func TestConnect(t *testing.T) {
	serverAddr := newTestServer(t)

	tests := []struct {
		desc          string
		dep           *rdpb.ResolvedDependency
		iface         string
		wantMetadata  map[string][]string
		wantErrorType error
		wantError     string
	}{
		{
			desc: "success",
			dep: &rdpb.ResolvedDependency{
				Interfaces: map[string]*rdpb.ResolvedDependency_Interface{
					"grpc://intrinsic_proto.assets.dependencies.testing.TestService": {
						Protocol: &rdpb.ResolvedDependency_Interface_Grpc_{
							Grpc: &rdpb.ResolvedDependency_Interface_Grpc{
								Connection: &gcpb.GrpcConnection{
									Address: serverAddr,
									Metadata: []*gcpb.GrpcConnection_Metadata{
										{
											Key:   "test_key",
											Value: "test_value1",
										},
										{
											Key:   "test_key",
											Value: "test_value2",
										},
									},
								},
							},
						},
					},
				},
			},
			iface: "grpc://intrinsic_proto.assets.dependencies.testing.TestService",
			wantMetadata: map[string][]string{
				"test_key": {"test_value1", "test_value2"},
			},
		},
		{
			desc:          "no interfaces",
			dep:           &rdpb.ResolvedDependency{},
			iface:         "grpc://intrinsic_proto.assets.dependencies.testing.TestService",
			wantErrorType: errMissingInterface,
			wantError:     "no interfaces provided",
		},
		{
			desc: "wrong interface type",
			dep: &rdpb.ResolvedDependency{
				Interfaces: map[string]*rdpb.ResolvedDependency_Interface{
					"data://google.protobuf.Empty": {
						Protocol: &rdpb.ResolvedDependency_Interface_Data_{
							Data: &rdpb.ResolvedDependency_Interface_Data{
								Id: &idpb.Id{Package: "ai.intrinsic", Name: "data_asset"},
							},
						},
					},
				},
			},
			iface:         "grpc://intrinsic_proto.assets.dependencies.testing.TestService",
			wantErrorType: errMissingInterface,
			wantError:     "got interfaces: data://google.protobuf.Empty",
		},
		{
			desc: "not a gRPC interface",
			dep: &rdpb.ResolvedDependency{
				Interfaces: map[string]*rdpb.ResolvedDependency_Interface{
					"data://google.protobuf.Empty": {
						Protocol: &rdpb.ResolvedDependency_Interface_Data_{
							Data: &rdpb.ResolvedDependency_Interface_Data{
								Id: &idpb.Id{Package: "ai.intrinsic", Name: "data_asset"},
							},
						},
					},
				},
			},
			iface:         "data://google.protobuf.Empty",
			wantErrorType: errNotGRPC,
			wantError:     "is not gRPC",
		},
	}

	for _, tc := range tests {
		t.Run(tc.desc, func(t *testing.T) {
			conn, ctx, err := Connect(context.Background(), tc.dep, tc.iface)
			if tc.wantErrorType != nil || tc.wantError != "" {
				if err == nil {
					t.Errorf("Connect() returned no error, want error (%q, %q)", tc.wantErrorType, tc.wantError)
				} else if tc.wantErrorType != nil && !errors.Is(err, tc.wantErrorType) {
					t.Errorf("Connect() returned error %q, want error type %q", err, tc.wantErrorType)
				} else if !strings.Contains(err.Error(), tc.wantError) {
					t.Errorf("Connect() returned error %q, want error %q", err, tc.wantError)
				}
			} else if err != nil {
				t.Errorf("Connect() returned an unexpected error: %v", err)
			} else {
				defer conn.Close()

				client := tsgrpcpb.NewTestServiceClient(conn)
				response, err := client.Test(ctx, &tspb.TestRequest{})
				if err != nil {
					t.Fatalf("Test() returned an unexpected error: %v", err)
				}
				for k, vs := range tc.wantMetadata {
					gotVs := response.GetContextMetadata()[k].GetValues()
					if diff := cmp.Diff(vs, gotVs); diff != "" {
						t.Errorf("Test() returned an unexpected diff for metadata key %q (-want +got): %v", k, diff)
					}
				}
			}
		})
	}
}

func TestGetDataPayload(t *testing.T) {
	payload := &emptypb.Empty{}
	payloadAny := prototestutil.MustWrapInAny(t, payload)
	da := &dapb.DataAsset{
		Data:              payloadAny,
		FileDescriptorSet: descriptor.FileDescriptorSetFrom(payload),
		Metadata: &mpb.Metadata{
			AssetType: atpb.AssetType_ASSET_TYPE_DATA,
			IdVersion: &idpb.IdVersion{
				Id: &idpb.Id{
					Package: "ai.intrinsic",
					Name:    "data_asset",
				},
			},
		},
	}

	tests := []struct {
		desc          string
		dep           *rdpb.ResolvedDependency
		iface         string
		wantPayload   *anypb.Any
		wantErrorType error
		wantError     string
	}{
		{
			desc: "success",
			dep: &rdpb.ResolvedDependency{
				Interfaces: map[string]*rdpb.ResolvedDependency_Interface{
					"data://google.protobuf.Empty": {
						Protocol: &rdpb.ResolvedDependency_Interface_Data_{
							Data: &rdpb.ResolvedDependency_Interface_Data{
								Id: &idpb.Id{Package: "ai.intrinsic", Name: "data_asset"},
							},
						},
					},
				},
			},
			iface:       "data://google.protobuf.Empty",
			wantPayload: payloadAny,
		},
		{
			desc:          "no interfaces",
			dep:           &rdpb.ResolvedDependency{},
			iface:         "data://google.protobuf.Empty",
			wantErrorType: errMissingInterface,
			wantError:     "no interfaces provided",
		},
		{
			desc: "wrong interface type",
			dep: &rdpb.ResolvedDependency{
				Interfaces: map[string]*rdpb.ResolvedDependency_Interface{
					"grpc://intrinsic_proto.assets.dependencies.testing.TestService": {
						Protocol: &rdpb.ResolvedDependency_Interface_Grpc_{
							Grpc: &rdpb.ResolvedDependency_Interface_Grpc{
								Connection: &gcpb.GrpcConnection{
									Address: "localhost:12345",
								},
							},
						},
					},
				},
			},
			iface:         "data://google.protobuf.Empty",
			wantErrorType: errMissingInterface,
			wantError:     "got interfaces: grpc://intrinsic_proto.assets.dependencies.testing.TestService",
		},
		{
			desc: "not a data interface",
			dep: &rdpb.ResolvedDependency{
				Interfaces: map[string]*rdpb.ResolvedDependency_Interface{
					"grpc://intrinsic_proto.assets.dependencies.testing.TestService": {
						Protocol: &rdpb.ResolvedDependency_Interface_Grpc_{
							Grpc: &rdpb.ResolvedDependency_Interface_Grpc{
								Connection: &gcpb.GrpcConnection{
									Address: "localhost:12345",
								},
							},
						},
					},
				},
			},
			iface:         "grpc://intrinsic_proto.assets.dependencies.testing.TestService",
			wantErrorType: errNotData,
			wantError:     "is not data",
		},
	}

	ctx := context.Background()
	fake := fakedataassets.StartServer(ctx, t, fakedataassets.WithDataAssets([]*dapb.DataAsset{da}))

	for _, tc := range tests {
		t.Run(tc.desc, func(t *testing.T) {
			gotPayload, err := GetDataPayload(ctx, tc.dep, tc.iface, WithDataAssetsClient(fake.Client))
			if tc.wantErrorType != nil || tc.wantError != "" {
				if err == nil {
					t.Errorf("GetDataPayload() returned no error, want error (%q, %q)", tc.wantErrorType, tc.wantError)
				} else if tc.wantErrorType != nil && !errors.Is(err, tc.wantErrorType) {
					t.Errorf("GetDataPayload() returned error %q, want error type %q", err, tc.wantErrorType)
				} else if !strings.Contains(err.Error(), tc.wantError) {
					t.Errorf("GetDataPayload() returned error %q, want error %q", err, tc.wantError)
				}
			} else if err != nil {
				t.Errorf("GetDataPayload() returned an unexpected error: %v", err)
			} else if diff := cmp.Diff(tc.wantPayload, gotPayload, protocmp.Transform()); diff != "" {
				t.Errorf("GetDataPayload() returned an unexpected diff (-want +got): %v", diff)
			}
		})
	}
}

func TestHasDependencies(t *testing.T) {
	tests := []struct {
		desc       string
		descriptor protoreflect.MessageDescriptor
		options    []ResolvedDepsIntrospectionOption
		want       bool
	}{
		{
			desc:       "simple_grpc_dependency_config",
			descriptor: (&tcpb.SimpleGrpcDependencyConfig{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "simple_grpc_dependency_config_with_dependency_annotation_check",
			descriptor: (&tcpb.SimpleGrpcDependencyConfig{}).ProtoReflect().Descriptor(),
			options:    []ResolvedDepsIntrospectionOption{WithDependencyAnnotation()},
			want:       true,
		},
		{
			desc:       "simple_grpc_dependency_config_with_dependency_annotation_and_skill_annotations_check",
			descriptor: (&tcpb.SimpleGrpcDependencyConfig{}).ProtoReflect().Descriptor(),
			options:    []ResolvedDepsIntrospectionOption{WithDependencyAnnotation(), WithSkillAnnotations()},
			want:       false,
		},
		{
			desc:       "simple_grpc_dependency_config_with_skill_annotations_check",
			descriptor: (&tcpb.SimpleGrpcDependencyConfig{}).ProtoReflect().Descriptor(),
			options:    []ResolvedDepsIntrospectionOption{WithSkillAnnotations()},
			want:       false,
		},
		{
			desc:       "simple_data_dependency_config",
			descriptor: (&tcpb.SimpleDataDependencyConfig{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "simple_object_dependency_config",
			descriptor: (&tcpb.SimpleObjectDependencyConfig{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_repeated_dependency",
			descriptor: (&tcpb.WithRepeatedDependency{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_nested_fields",
			descriptor: (&tcpb.WithNestedFields{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_multiple_dependencies",
			descriptor: (&tcpb.WithMultipleDependencies{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_unannotated_dependency",
			descriptor: (&tcpb.WithUnannotatedDependency{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_incorrect_dependency_type",
			descriptor: (&tcpb.WithIncorrectMessageType{}).ProtoReflect().Descriptor(),
			want:       false,
		},
		{
			desc:       "config_with_multiple_interfaces",
			descriptor: (&tcpb.WithMultipleInterfaces{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_map_dependency",
			descriptor: (&tcpb.WithMapDependency{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_optional_dependency",
			descriptor: (&tcpb.WithOptionalDependency{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_required_dependency",
			descriptor: (&tcpb.WithRequiredDependency{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "config_with_any_proto",
			descriptor: (&tcpb.WithAnyProto{}).ProtoReflect().Descriptor(),
			want:       false,
		},
		{
			desc:       "skill_with_annotations_dependency_check",
			descriptor: (&tcpb.WithAlwaysProvideConnectionInfo{}).ProtoReflect().Descriptor(),
			want:       true,
		},
		{
			desc:       "skill_with_annotations_only_dependency_annotation_check",
			descriptor: (&tcpb.WithAlwaysProvideConnectionInfo{}).ProtoReflect().Descriptor(),
			options:    []ResolvedDepsIntrospectionOption{WithDependencyAnnotation()},
			want:       true,
		},
		{
			desc:       "skill_with_annotations_skill_annotation_check",
			descriptor: (&tcpb.WithAlwaysProvideConnectionInfo{}).ProtoReflect().Descriptor(),
			options:    []ResolvedDepsIntrospectionOption{WithDependencyAnnotation(), WithSkillAnnotations()},
			want:       true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.desc, func(t *testing.T) {
			got := HasResolvedDependency(tc.descriptor, (tc.options)...)
			if got != tc.want {
				t.Errorf("HasResolvedDependency(%v) = %v, want: %v", tc.descriptor, got, tc.want)
			}
		})
	}
}

File: intrinsic/assets/dependencies/utils_test.py
# Copyright 2023 Intrinsic Innovation LLC

"""Tests for the Asset dependencies utility functions."""

import dataclasses
from typing import cast

from absl.testing import absltest
from absl.testing import parameterized
from google.protobuf import any_pb2
from google.protobuf import descriptor_pb2
from google.protobuf import empty_pb2
import grpc
from grpc.framework.foundation import logging_pool
from intrinsic.assets.data import fake_data_assets
from intrinsic.assets.data.proto.v1 import data_asset_pb2
from intrinsic.assets.dependencies import utils
from intrinsic.assets.dependencies.testing import test_service_pb2
from intrinsic.assets.dependencies.testing import test_service_pb2_grpc
from intrinsic.assets.proto import asset_type_pb2
from intrinsic.assets.proto import id_pb2
from intrinsic.assets.proto import metadata_pb2
from intrinsic.assets.proto.v1 import resolved_dependency_pb2


def _make_empty_data_asset(
    name: str = "data_asset",
) -> data_asset_pb2.DataAsset:
  payload = empty_pb2.Empty()
  payload_any = any_pb2.Any()
  payload_any.Pack(payload)

  return data_asset_pb2.DataAsset(
      data=payload_any,
      file_descriptor_set=descriptor_pb2.FileDescriptorSet(),
      metadata=metadata_pb2.Metadata(
          asset_type=asset_type_pb2.AssetType.ASSET_TYPE_DATA,
          id_version=id_pb2.IdVersion(
              id=id_pb2.Id(package="ai.intrinsic", name=name),
              version="0.0.1",
          ),
      ),
  )


@dataclasses.dataclass(frozen=True)
class ConnectTestCase:
  desc: str
  dep: resolved_dependency_pb2.ResolvedDependency
  iface: str
  want_metadata: dict[str, list[str]] | None = None
  want_error: type[BaseException] | None = None
  want_error_message_regex: str | None = None


@dataclasses.dataclass(frozen=True)
class GetDataPayloadTestCase:
  desc: str
  dep: resolved_dependency_pb2.ResolvedDependency
  iface: str
  want_payload: any_pb2.Any | None = None
  want_error: type[BaseException] | None = None
  want_error_message_regex: str | None = None


_CONNECT_TEST_CASES = [
    ConnectTestCase(
        desc="success",
        dep=resolved_dependency_pb2.ResolvedDependency(
            interfaces={
                "grpc://intrinsic_proto.assets.dependencies.testing.TestService": resolved_dependency_pb2.ResolvedDependency.Interface(
                    grpc_connection=resolved_dependency_pb2.ResolvedDependency.Interface.GrpcConnection(
                        address="localhost:12345",
                        metadata=[
                            resolved_dependency_pb2.ResolvedDependency.Interface.GrpcConnection.Metadata(
                                key="test_key",
                                value="test_value1",
                            ),
                            resolved_dependency_pb2.ResolvedDependency.Interface.GrpcConnection.Metadata(
                                key="test_key",
                                value="test_value2",
                            ),
                        ],
                    ),
                ),
            },
        ),
        iface="grpc://intrinsic_proto.assets.dependencies.testing.TestService",
        want_metadata={
            "test_key": ["test_value1", "test_value2"],
        },
    ),
    ConnectTestCase(
        desc="no interfaces",
        dep=resolved_dependency_pb2.ResolvedDependency(),
        iface="grpc://intrinsic_proto.assets.dependencies.testing.TestService",
        want_error=utils.MissingInterfaceError,
    ),
    ConnectTestCase(
        desc="wrong interface type",
        dep=resolved_dependency_pb2.ResolvedDependency(
            interfaces={
                "data://google.protobuf.Empty": resolved_dependency_pb2.ResolvedDependency.Interface(
                    data=resolved_dependency_pb2.ResolvedDependency.Interface.Data(
                        id=id_pb2.Id(package="ai.intrinsic", name="data_asset"),
                    ),
                ),
            },
        ),
        iface="grpc://intrinsic_proto.assets.dependencies.testing.TestService",
        want_error=utils.MissingInterfaceError,
        want_error_message_regex="got interfaces: data://google.protobuf.Empty",
    ),
    ConnectTestCase(
        desc="not a gRPC connection",
        dep=resolved_dependency_pb2.ResolvedDependency(
            interfaces={
                "data://google.protobuf.Empty": resolved_dependency_pb2.ResolvedDependency.Interface(
                    data=resolved_dependency_pb2.ResolvedDependency.Interface.Data(
                        id=id_pb2.Id(package="ai.intrinsic", name="data_asset"),
                    ),
                ),
            },
        ),
        iface="data://google.protobuf.Empty",
        want_error=utils.NotGRPCError,
    ),
]

_GET_DATA_PAYLOAD_TEST_CASES = [
    GetDataPayloadTestCase(
        desc="success",
        dep=resolved_dependency_pb2.ResolvedDependency(
            interfaces={
                "data://google.protobuf.Empty": resolved_dependency_pb2.ResolvedDependency.Interface(
                    data=resolved_dependency_pb2.ResolvedDependency.Interface.Data(
                        id=id_pb2.Id(package="ai.intrinsic", name="data_asset"),
                    ),
                ),
            },
        ),
        iface="data://google.protobuf.Empty",
        want_payload=_make_empty_data_asset().data,
    ),
    GetDataPayloadTestCase(
        desc="no interfaces",
        dep=resolved_dependency_pb2.ResolvedDependency(),
        iface="data://google.protobuf.Empty",
        want_error=utils.MissingInterfaceError,
        want_error_message_regex="no interfaces provided",
    ),
    GetDataPayloadTestCase(
        desc="wrong interface type",
        dep=resolved_dependency_pb2.ResolvedDependency(
            interfaces={
                "grpc://intrinsic_proto.assets.dependencies.testing.TestService": resolved_dependency_pb2.ResolvedDependency.Interface(
                    grpc_connection=resolved_dependency_pb2.ResolvedDependency.Interface.GrpcConnection(
                        address="localhost:12345",
                    ),
                ),
            },
        ),
        iface="data://google.protobuf.Empty",
        want_error=utils.MissingInterfaceError,
        want_error_message_regex=(
            "got interfaces:"
            " grpc://intrinsic_proto.assets.dependencies.testing.TestService"
        ),
    ),
    GetDataPayloadTestCase(
        desc="not data",
        dep=resolved_dependency_pb2.ResolvedDependency(
            interfaces={
                "grpc://intrinsic_proto.assets.dependencies.testing.TestService": resolved_dependency_pb2.ResolvedDependency.Interface(
                    grpc_connection=resolved_dependency_pb2.ResolvedDependency.Interface.GrpcConnection(
                        address="localhost:12345",
                    ),
                ),
            },
        ),
        iface="grpc://intrinsic_proto.assets.dependencies.testing.TestService",
        want_error=utils.NotDataError,
    ),
]


class TestService(test_service_pb2_grpc.TestServiceServicer):

  def Test(
      self,
      request: test_service_pb2.TestRequest,
      context: grpc.ServicerContext,
  ) -> test_service_pb2.TestResponse:
    response = test_service_pb2.TestResponse(
        context_metadata={},
    )
    for k, vs in context.invocation_metadata():
      response.context_metadata[k].values.append(vs)
    return response


class UtilsTest(parameterized.TestCase):

  @parameterized.parameters(*_CONNECT_TEST_CASES)
  def test_connect(self, tc: ConnectTestCase):
    if tc.want_error is not None:
      with self.assertRaisesRegex(tc.want_error, tc.want_error_message_regex):
        utils.connect(tc.dep, tc.iface)
    else:
      server = grpc.server(
          logging_pool.pool(max_workers=1),
          options=[("grpc.max_receive_message_length", -1)],
      )
      test_service_pb2_grpc.add_TestServiceServicer_to_server(
          TestService(), server
      )
      server.add_insecure_port(
          tc.dep.interfaces[tc.iface].grpc_connection.address
      )
      server.start()

      channel = utils.connect(tc.dep, tc.iface)
      stub = test_service_pb2_grpc.TestServiceStub(channel)
      response = stub.Test(test_service_pb2.TestRequest())
      channel.close()
      server.stop(None)
      for k, vs in cast(dict[str, list[str]], tc.want_metadata).items():
        self.assertIn(k, response.context_metadata)
        self.assertCountEqual(vs, response.context_metadata[k].values)

  @parameterized.parameters(*_GET_DATA_PAYLOAD_TEST_CASES)
  def test_get_data_payload(self, tc: GetDataPayloadTestCase):
    stub, cleanup = fake_data_assets.FakeDataAssetsService.start_server(
        [_make_empty_data_asset()]
    )

    try:
      if tc.want_error is not None:
        with self.assertRaisesRegex(tc.want_error, tc.want_error_message_regex):
          utils.get_data_payload(
              dep=tc.dep, iface=tc.iface, data_assets_client=stub
          )
      else:
        got_payload = utils.get_data_payload(
            dep=tc.dep, iface=tc.iface, data_assets_client=stub
        )
        self.assertEqual(got_payload, tc.want_payload, msg=tc.desc)
    finally:
      cleanup()


if __name__ == "__main__":
  absltest.main()
